src/MoBi.Presentation/DTO/MoleculeSelectionDTO.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using MoBi.Assets;
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Presentation.DTO;
+using OSPSuite.Utility.Validation;
+
+namespace MoBi.Presentation.DTO
+{
+   public class MoleculeSelectionDTO : DxValidatableDTO
+   {
+      public MoleculeSelectionDTO()
+      {
+         Rules.Add(AllRules.SelectedMoleculesHaveUniqueNames);
+      }
+
+      private Func<IReadOnlyList<MoleculeSelectionDTO>> _getSelectedMolecules;
+
+      public string BuildingBlock { get; set; }
+
+      public string MoleculeName => MoleculeBuilder?.Name;
+
+      public MoleculeBuilder MoleculeBuilder { get; set; }
+
+      public bool Selected { get; set; }
+
+      public string Icon => MoleculeBuilder.Icon;
+
+      public void AddSelectedMoleculeRetriever(Func<IReadOnlyList<MoleculeSelectionDTO>> getSelectedMolecules)
+      {
+         _getSelectedMolecules = getSelectedMolecules;
```

Comment:
@msevestre my question is here. I need to invalidate at the selected molecule level so that the grid shows messages on the appropriate rows.

So every MoleculeSelectionDTO gets this retriever that calculates the already selected molecules, which is used to check for other selections with the same name.

I can't pass that as a list, because it will be calculated at time of construction of the object instead of at time of validation.

Do you think there's some other way to invalidate this object if another object is selected that has the same name?